va/va_prot.h

```diff
@@ -0,0 +1,361 @@
+    uint32_t                        status;
+    uint32_t                        va_reserved[VA_PADDING_LOW];
+} VAProtectedSessionTeeExecBuffer;
+
+/** \brief Buffer for vaProtectedSessionHwUpdate() */
+typedef struct _VAProtectedSessionHwUpdateBuffer {
+    /** \brief Input buffer of VAProtectedFuncionBufferInput structure */
+    VAProtectedFuncionBufferInput   input;
+    /** \brief Output buffer of VAProtectedFuncionBufferOutput structure */
+    VAProtectedFuncionBufferOutput  output;
+    /** \brief Return the result of this function. It is IHV's implementation specific */
+    uint32_t                        status;
+    uint32_t                        va_reserved[VA_PADDING_LOW];
+} VAProtectedSessionHwUpdateBuffer;
+
+/** \brief Protected Link TeeExec Function Codes. */
+typedef enum _VA_TEE_EXEC_PROTECTED_LINK_FUNCTION_ID
+{
+    // 0x20000000~0x200000FFF reserved for TEE Exec Protected Link function
+    VA_TEE_EXEC_PROTECTED_LINK_FUNCID_SET_PROTECTION_LEVEL  = 0x20000000,
+
+} VA_TEE_EXEC_PROTECTED_LINK_FUNCTION_ID;
+
+/**
+ * \brief Create a protected session
+ *
+ * Create a protected session
+ *
+ * @param[in] dpy                   the VA display
+ * @param[in] config_id             configuration for the protected session
+ * @param[out] protected_session    created protected session id upon return
+ */
+VAStatus vaCreateProtectedSession(
+    VADisplay dpy,
+    VAConfigID config_id,
+    VAProtectedSessionID *protected_session
+);
+
+/**
+ * \brief Destroy a protected session
+ *
+ * Destroy a protected session
+ *
+ * @param[in] dpy                   the VA display
+ * @param[in] protected_session     protected session to be destroyed
+ */
+VAStatus vaDestroyProtectedSession(
+    VADisplay dpy,
+    VAProtectedSessionID protected_session
+);
+
+/**
+ * \brief Attach a protected session to the display
+ *
+ * Attach a protected session to the display to enable protection
+ *
+ * @param[in] dpy                   the VA display
+ * @param[in] reserved              reserved for future usage
+ * @param[in] protected_session     the protected session to attach
+ */
+VAStatus vaAttachProtectedSession(
+    VADisplay dpy,
+    VAContextID reserved,
+    VAProtectedSessionID protected_session
+);
+
+/**
+ * \brief Detach the protected session of the display
+ *
+ * Detach protected session of the display to disable protection
+ *
+ * @param[in] dpy                   the VA display
+ * @param[in] reserved              reserved for future usage
+ */
+VAStatus vaDetachProtectedSession(
+    VADisplay dpy,
+    VAContextID reserved
+);
+
+/**
+ * \brief TEE execution for the particular protected session
+ *
+ * TEE execution for the particular protected session
+ *
+ * @param[in] dpy                   the VA display
+ * @param[in] protected_session     the protected session
+ * @param[in,out] buf_id            the VA buffer
+ */
+VAStatus vaProtectedSessionTeeExec(
```

Comment:
Is this a blocking operation? I.e. after exiting from this function is it guaranteed that operation is completed? Actually VAAPI is async API, we can use vaSyncBuffer() for this purpose...

Reply:
In some cases, it retrieves data from TEE which is a blocking operation.